pkg/urbit/vere/lmdb.c

```diff
@@ -0,0 +1,547 @@
+/* vere/lmdb.c
+*/
+
+#include "all.h"
+
+#include <uv.h>
+#include <lmdb.h>
+
+#include "vere/vere.h"
+
+// Event log persistence for Urbit
+//
+// Persistence works by having an lmdb environment opened on the main
+// thread. This environment is used to create read-only transactions
+// synchronously when needed.
+//
+// But the majority of lmdb writes operate asynchronously in the uv worker
+// pool. Since individual transactions are bound to threads, we perform all
+// blocking writing on worker threads.
+//
+// We perform the very first metadata writes on the main thread because we
+// can't do anything until they persist.
+
+/* u3_lmdb_init(): Opens up a log environment
+**
+** Precondition: log_path points to an already created directory
+*/
+MDB_env* u3_lmdb_init(const char* log_path)
+{
+  MDB_env* env = 0;
+  c3_w ret_w = mdb_env_create(&env);
+  if (ret_w != 0) {
+    u3l_log("lmdb: init fail: %s\n", mdb_strerror(ret_w));
+    return 0;
+  }
+
+  // Our databases have up to three tables: META, EVENTS, and GRAINS.
+  ret_w = mdb_env_set_maxdbs(env, 3);
+  if (ret_w != 0) {
+    u3l_log("lmdb: failed to set number of databases: %s\n", mdb_strerror(ret_w));
+    return 0;
+  }
+
+  // TODO: Start with a gigabyte for the event log.
+  //
+  ret_w = mdb_env_set_mapsize(env, 1024 * 1024 * 1024);
+  if (ret_w != 0) {
+    u3l_log("lmdb: failed to set database size: %s\n", mdb_strerror(ret_w));
+    return 0;
+  }
+
+  ret_w = mdb_env_open(env, log_path, 0, 0664);
+  if (ret_w != 0) {
+    u3l_log("lmdb: failed to open event log: %s\n", mdb_strerror(ret_w));
+    return 0;
+  }
+
+  return env;
+}
+
+/* u3_lmdb_shutdown(): Shuts down lmdb
+*/
+void u3_lmdb_shutdown(MDB_env* env)
+{
+  mdb_env_close(env);
+}
+
+/* _perform_put_on_databse_raw(): Writes a key/value pair to a specific
+** database as part of a transaction.
+**
+** The raw version doesn't take ownership of either key/value and performs no
+** nock calculations, so it is safe to call from any thread.
+*/
+static
+void _perform_put_on_databse_raw(MDB_txn* transaction_u,
+                                 MDB_dbi database_u,
+                                 c3_w flags,
+                                 void* key,
+                                 size_t key_len,
+                                 void* value,
+                                 size_t value_len) {
+  MDB_val key_val, value_val;
+
+  key_val.mv_size = key_len;
+  key_val.mv_data = key;
+
+  value_val.mv_size = value_len;
+  value_val.mv_data = value;
+
+  c3_w ret_w = mdb_put(transaction_u, database_u, &key_val, &value_val, flags);
+  if (ret_w != 0) {
+    u3l_log("lmdb: write failed: %s\n", mdb_strerror(ret_w));
+    u3m_bail(c3__fail);
```

Comment:
`u3m_bail` implements the nock runtime error handling protocol, which isn't a good fit here. For synchronous functions like this and the read functions, putting the error status in the return makes sense. For the top-level asynchronous writes, we'll either need to add an error parameter to the `on_complete()` callback, or have separate success and failure callbacks. It looks like the `uv_write_t` can be explicitly cancelled with `uv_cancel()`, and the cancellation can be observed in the after_cb. In `pier.c`, on read/write errors, we should just `u3_pier_bail()`.